package ru.practicum.shareit.exceptions;

import jakarta.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

@Slf4j
@RestControllerAdvice
public class ExceptionsHandler {

    @ExceptionHandler
    @ResponseStatus(HttpStatus.NOT_FOUND)
    public ErrorResponse handlerNotFoundException(NotFoundException e) {
        log.debug("Произошла ошибка NotFoundException. message: {}, Trace: {}", e.getMessage(), e.getStackTrace());
        return new ErrorResponse(e.getMessage());
    }

    @ExceptionHandler
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ErrorResponse handlerValidationException(ValidationException e) {
        log.debug("Произошла ошибка ValidationException. message: {}, Trace: {}", e.getMessage(), e.getStackTrace());
        return new ErrorResponse(e.getMessage());
    }

    @ExceptionHandler
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    public ErrorResponse handlerThrowable(Throwable e) {
        log.debug("Произошла ошибка handlerThrowable. message: {}, Trace: {}", e.getMessage(), e.getStackTrace());
        return new ErrorResponse(e.getMessage());
    }

    //Проверка @Validated
    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    public ErrorResponse handlerAnnootation(MethodArgumentNotValidException e) {
        log.debug("Произошла ошибка валидации MethodArgumentNotValidException. message: {}", e.getMessage());
        return new ErrorResponse(e.getMessage());
    }

    @ExceptionHandler
    public ResponseEntity<String> handlerValidAnnotation(ConstraintViolationException e) {
        log.debug("Произошла ошибка валидации ConstraintViolationException. message: {}", e.getMessage());
        return new ResponseEntity<>(e.getMessage(), HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler
    @ResponseStatus(HttpStatus.CONFLICT)
    public ErrorResponse dublicateValidationException(DublicateException e) {
        log.debug("Произошла ошибка DublicateException. message: {}, Trace: {}", e.getMessage(), e.getStackTrace());
        return new ErrorResponse(e.getMessage());
    }
}